### PR TITLE
feat(distributed): RPC-based distributed training support and add distributed MAML example

### LINKED ISSUE #57: [Feature Request] Task-level Optimization with Distributed Data Parallelization
## Motivation

Task-level parallelization for multi-host multi-process optimization.

Batch-level parallelization can be implemented easily by wrapping the network (`nn.Module`) with:

- [`torch.nn.DataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html) (single-host multi-GPUs) (SPMD)
- [`torch.nn.parallel.DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) (multi-host multi-GPUs)

However, for algorithms that require task-level parallelization, non of the above solutions work. `torch.nn.DataParallel` and `torch.nn.parallel.DistributedDataParallel` are used for module-level parallelization. The wrapper will replicate the user module to multiple copies, then do the forward pass in parallel. For task-level parallelization, each task needs to maintain its own model parameters and (optional) training data. The module parameters may be different across tasks.

## Solution

`functorch.vmap` + distributed data parallel optimization.

## Example

```python
import torch
import torch.distributed.autograd as dist_autograd
import torch.distributed.rpc as rpc
from torch import optim
from torch.distributed.optim import DistributedOptimizer

with dist_autograd.context() as context_id:
    # Forward pass.
    rref1 = rpc.remote("worker1", torch.add, args=(torch.ones(2), 3))
    rref2 = rpc.remote("worker1", torch.add, args=(torch.ones(2), 1))
    loss = rref1.to_here() + rref2.to_here()
 
    # Backward pass.
    dist_autograd.backward(context_id, [loss.sum()])
 
    # Optimizer.
    dist_optim = DistributedOptimizer(
        optim.SGD,
        [rref1, rref2],
        lr=0.05,
    )
    dist_optim.step(context_id)
```

## Additional context

**Resources:**

PyTorch:

- Tutorial: [PyTorch Distributed Overview](https://pytorch.org/tutorials/beginner/dist_overview.html#)
- Tutorial: [Distributed Data Parallel](https://pytorch.org/docs/stable/notes/ddp.html#)
- API: [Module level Data Parallel `torch.nn.DataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html) (SPMD)
- API: [Module level Distributed Data Parallel `torch.nn.parallel.DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html)
- API: [PyTorch Distributed Optimizers `torch.distributed.optim`](https://pytorch.org/docs/stable/distributed.optim.html)
- API: [Vectorization map `functorch.vmap`](https://pytorch.org/functorch/stable/generated/functorch.vmap.html)
- API: [NVIDIA apex.parallel](https://nvidia.github.io/apex/)

JAX:

- Tutorial: [Named axes and easy-to-revise parallelism](https://jax.readthedocs.io/en/latest/notebooks/xmap_tutorial.html)
- API: [Vectorization map `jax.vmap`](https://jax.readthedocs.io/en/latest/jax.html?highlight=pmap#vectorization-vmap)
- API: [Parallel map `jax.pmap`](https://jax.readthedocs.io/en/latest/jax.html?highlight=pmap#parallelization-pmap) (SPMD)
- API _(Experimental)_: [`jax.experimental.maps.xmap`](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.maps.xmap.html#jax-experimental-maps-xmap)
- Tutorial: [Using JAX in multi-host and multi-process environments](https://jax.readthedocs.io/en/latest/multi_process.html)

## Checklist

- [X] I have checked that there is no similar issue in the repo (**required**)

### PR DESCRIPTION
**Pull request by XuehaiPan**
## Description

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [X] I have raised an issue to propose this change ([required](https://github.com/metaopt/torchopt/issues) for new features and bug fixes)

Resolves #57

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

### New APIs

- `torchopt.distributed.is_available`
- `torchopt.distributed.backward`

World and Process Group:

- `torchopt.distributed.auto_init_rpc`
- `torchopt.distributed.get_world_info`
- `torchopt.distributed.get_world_rank` (`torchopt.distributed.get_rank`)
- `torchopt.distributed.get_world_size`
- `torchopt.distributed.get_local_rank`
- `torchopt.distributed.get_local_world_size`
- `torchopt.distributed.barrier`

Wrappers:

- `torchopt.distributed.auto_init_rpc`
- `torchopt.distributed.on_rank`
- `torchopt.distributed.not_on_rank`
- `torchopt.distributed.rank_zero_only`
- `torchopt.distributed.rank_non_zero_only`

Remote call:

- `torchopt.distributed.remote_async_call`
- `torchopt.distributed.remote_sync_call`
- `torchopt.distributed.parallelize` (`torchopt.distributed.parallelize_sync`)
- `torchopt.distributed.parallelize_async`

Misc:

- `torchopt.distributed.dim_partitioner`
- `torchopt.distributed.batch_partitioner`
- `torchopt.distributed.exclusive_batch_partitioner`
- `torchopt.distributed.mean_reducer`
- `torchopt.distributed.sum_reducer`

Examples:

```python
import torchopt.distributed as todist

def worker_init_fn():
    # set process title, seeding, etc.

@todist.auto_init_rpc(worker_init_fn)
def main():
    ...
	model = Model(...)
	train(model)        # execute on rank 0 only
    save_model(model)   # execute on rank 0 only
```

```python
@todist.rank_zero_only
def save_model(model):
    ...

@todist.rank_zero_only
def train(model):
    model_rref = todist.rpc.RRef(model_rref)
    dataloader = DataLoader(...)
    optimizer = Optimizer(...)

    for batch in dataloader:
        optimizer.zero_grad()
        with todist.autograd.context() as context_id:
            loss = compute_loss(model_rref, batch)
            todist.autograd.backward(context_id, loss)  # in single process, we use `loss.backward()`
            optimizer.step()
```

```python
@todist.parallelize(partitioner=todist.batch_partitioner, reducer=todist.mean_reducer)
def compute_loss(model_rref, batch):
    model = model_rref.to_here()
    loss = ...
    return loss
```

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://torchopt.readthedocs.io/en/latest/developer/contributing.html) guide (**required**)
- [X] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [X] I have reformatted the code using `make format` (**required**)
- [X] I have checked the code using `make lint` (**required**)
- [X] I have ensured `make test` pass. (**required**)

### CODE REVIEW

File: torchopt/distributed/autograd.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        if inputs is not None:
+            inputs = set(inputs)  # type: ignore[assignment]
+            all_local_grads = {p: g for p, g in all_local_grads.items() if p in inputs}
+
+        with LOCK:
+            for p, g in all_local_grads.items():
+                if retain_graph and getattr(p, 'grad', None) is not None:
+                    p.grad = p.grad.add(g)
+                else:
+                    p.grad = g
```
Comment (XuehaiPan) [inline torchopt/distributed/autograd.py:84]: @JieRen98 Does this correct for `retain_graph` and `accumulate_grad=True` in [`torch.autograd.backward`](https://pytorch.org/docs/stable/_modules/torch/autograd.html#backward)? Or maybe just `p.grad = g`.

File: torchopt/distributed/autograd.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        for p in inputs:
+            try:
+                grads.append(all_local_grads[p])
+            except KeyError as ex:
+                if not allow_unused:
+                    raise RuntimeError(
+                        'One of the differentiated Tensors appears to not have been used in the '
+                        'graph. Set allow_unused=True if this is the desired behavior.'
+                    ) from ex
+                grads.append(torch.zeros_like(p))
```
Comment (JieRen98) [inline torchopt/distributed/autograd.py:125]: 应该是append(None)吧？

File: torchopt/distributed/autograd.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        roots = [tensors] if isinstance(tensors, torch.Tensor) else list(tensors)
+        autograd.backward(autograd_ctx_id, roots=roots, retain_graph=retain_graph)
+
+        all_local_grads = autograd.get_gradients(autograd_ctx_id)
+        if inputs is not None:
+            inputs = set(inputs)  # type: ignore[assignment]
+            all_local_grads = {p: g for p, g in all_local_grads.items() if p in inputs}
+
+        with LOCK:
```
Comment (JieRen98) [inline torchopt/distributed/autograd.py:79]: 这个锁是什么情景会有用？

File: torchopt/linear_solve.py [hunk 4]
```diff
@@ -57,12 +60,12 @@ def ridge_matvec(v: TensorTree) -> TensorTree:
 
 
 def solve_cg(
-    matvec: Callable[[TensorTree], TensorTree],
-    b: TensorTree,
+    matvec: Callable[['TensorTree'], 'TensorTree'],
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torchopt/linear_solve.py]: `ForwardRef`

docstyle 不识别另一个包里定义的 type hint (`PyTree` from `optree.typing`)。

File: torchopt/distributed/autograd.py (hunk 3 above)
Comment (XuehaiPan) [reply torchopt/distributed/autograd.py:79]: 同一个 worker 有好几个 thread 同时做 backward.

Comment (JieRen98) [reply torchopt/distributed/autograd.py:79]: 有没有试过直接不用锁会不会内存竞争？

Comment (XuehaiPan) [reply torchopt/distributed/autograd.py:79]: 只是用户可能会出现这样的情况，而且 PyTorch 的 distributed example 推荐加锁。如果就一个 thread，不会出现竞争锁的问题。

**Pull request merged**